### ISSUE TITLE
Order independence of enabling instrumentation vs importing instrumented libraries

### ISSUE THREAD

**Issue by anuraaga** [open-telemetry/opentelemetry-ruby#1017]
I noticed that it seems there is an ordering dependency between enabling instrumentation and requiring instrumented libraries. It seems the libraries themselves have to be required before instrumentation can be applied.

https://github.com/anuraaga/sinatra-hello-world/blob/main/app-manual/app.rb#L1

I wonder if there isn't a way to remove this ordering constraint. I rarely use Ruby, and my fumbling didn't find any good leads, other languages allow intercepting the loading of a module / class, but I couldn't find anything similar for `require`. Hoping there may be one though that the ruby masters know about.

The context is I am working on autoinstrumentation of apps with OpenTelemetry without any code changes. For example, just add an annotation to a k8s pod and it gets instrumentation, all managed by the infrastructure without apps having to worry, even in Dockerfiles. This has worked well for Java, Python, and Node. For reference, the NodeJS approach which is closest to what I've been trying with Ruby

https://github.com/open-telemetry/opentelemetry-operator/tree/main/autoinstrumentation/nodejs
`NODE_OPTIONS=--require autoinstrumentation.js`

I am pretty close to having logic running automatically before a Ruby app using similar mechanisms. Because Ruby doesn't search paths similar to node's node_modules automatic lookup, I had to keep a manifest of required deps and add them myself but that is fine I think.

Overall README at https://github.com/anuraaga/sinatra-hello-world
Key file: https://github.com/anuraaga/sinatra-hello-world/blob/main/autoinstrumentation/opentelemetry-instrument.rb

Unfortunately, unlike Java, Python, Node, where instrumentation can be applied before the user imports the logic modules, it seems that that's not the case for Ruby. If I import use `Bundler.require`, it does import `sinatra` and instrumentation does succeed to apply, but then the app just doesn't load for some reason. Anyways this seemed too hacky and I didn't dig further than that, ideally the ordering constraint itself could be removed somehow.

Does anyone have any ideas if there is some magic that can be used to allow applying instrumentations before libraries are imported? I'm not a Ruby user myself but would love to provide Ruby users autoinstrumentation!

**Comment by arielvalentin** [comment]
Thanks for sharing that specific examples with your repository. 

I'll take a closer look next week but `bundle exec` should be calling `bundle/setup` to configure the load paths so you shouldn't have to manually set the load paths. 

If you want to run your app without `bundle exec` then add `require "bundler/setup"` or invoke it `Bundler.setup` before running `Bundler.require`.

https://bundler.io/v2.2/guides/groups.html#how-to-manage-groups-of-gems

Have you tried the Sinatra example in this repository? Does that run for you?

**Comment by anuraaga** [comment]
Thanks @arielvalentin - the repository also has the `app-manual` folder which is a Sinatra app instrumented in the normal way for comparison and works fine. My objective is not to get an instrumented Sinatra app, but is to allow infrastructure-driven auto instrumentation of an app without any code changes in the app.

So regards bundler, the user is already using bundler with their own Gemfile - but I need to add search paths to OpenTelemetry modules which that bundle is not aware of. Ideally, bundler had  flag like `--additional-bundle=path` but I think the approach I am using is ok though and not too worried about it.

The main issue is just that I can't enable instrumentation before requiring the instrumented library, which I need for the approach for auto instrumentation. Hopefully going through the repo can give more clarity on what this is about.

**Comment by arielvalentin** [comment]
I'm still not entirely clear on your goals here but it is indeed the case that the libraries _must_ be required _before_ the instrumentation is programmatically _initialized_. This step _must_ occur _before_ the application is started. 

This is because most instrumentation is implemented as monkey patches or hooks that are prepended to the target libraries before the application is initialized.

Off of the top of my head I can't think of simple way of you adding instrumentation without changing the init script. 

🤔 If your customer is providing you with a container, then you'd probably need to have a base image with the gems installed globally, but then bundler would ignore them but you'd still need to require the gems and then at a minimum call the "configure" method on the SDK. 

That in and of itself necessitates a code change because this version of the SDK does not include a minimum  configuration script. 

💡 I wonder if the SDK could provide such a script using the environment variables as a default.

**Comment by anuraaga** [comment]
The main intent of the issue is to get the experts (which I'm definitely not :) ) together to see if there may be an alternative mechanism for applying instrumentation that could work without the ordering restriction. The motivation is to allow apps to have telemetry with no knowledge of OTel at all - a base image or script would unfortunately defeat that I think as then the app still needs to use said mechanism, it's not completely free.

If someone can help with getting this line from that linked readme to function it would be awesome, including considering alternate monkey patching approaches in instrumentation.

`RUBYLIB=../autoinstrumentation RUBYOPT="-ropentelemetry-instrument"  bundle exec ruby app.rb`

If there is absolutely no way, it is what it is. But we got it for Java, Python, and Node - somehow I expected Ruby to be at least as dynamic so would also give us a path to true auto-instrumentation in some way. I'm still hoping, but it sounds like I shouldn't be too optimistic.

**Comment by anuraaga** [comment]
I played with trying to apply instrumentation in response to a library being required and came up with a pattern that sort of works

https://github.com/anuraaga/sinatra-hello-world/blob/main/autoinstrumentation/opentelemetry-instrument.rb#L5

With this, the autoinstrumentation works and traces get exported without the user's app needing any changes!

It currently tries enabling the SDK twice, presumably the `self` and `name` matching could be improved to help that. But I am wondering if an approach like this has ever been considered. Instrumentation could first check if method to instrument already exists, in which case monkey patch, and otherwise add a method_added hook to do it later. Again, I don't really know Ruby and don't know how reasonable / scalable this approach is so hoping to get some ideas from the experts. Naturally we would have liked to have a callback after `require`, but I couldn't find any such mechanism but hope that I'm just missing it.

I think independent of the issue of autoinstrumentation (which I think can be a gamechanger for adoption though), it is a nice user experience improvement to now have to worry about ordering here so if there is a reasonable approach to take to apply to the instrumentations, hopefully it could be considered.

**Comment by ericmustin** [comment]
@anuraaga 

here's my two cents:
- The canonical way to do this in ruby is supported via rails, using a `railtie`. You can review many of the popular tracing gems in ruby out there for implementations of this approach, here's the docs for the one I wrote at my previous employer that you can review https://github.com/DataDog/dd-trace-rb/blob/master/docs/GettingStarted.md#automatic-instrumentation.

- I don't think this pattern is as well supported in other ruby frameworks, like sinatra, and implementations like what's been suggested are pretty incomplete imo (how would this work if a user is just using an http client instrumentation, or sidekiq, etc, we'd have to maintain these patches everywhere in every instrumentation, and how would we be able to have this feature play nicely with users who having existing configuration, or users who want to modify configuration).

- Is there specification around this? an OTEP? I'm dubious on the benefits of taking on the support burden for this feature living within the SDK without some more robust specification for it. In practice, I've anecdotally seen this feature has a double-edge sword effect of helping adoption and also ballooning the amount of support. In practice, by encouraging adoption largely among folks who, as you put it, wish to have 

> telemetry with no knowledge of OTel at all

You very quickly find that you are suddenly doing enterprise support for people with the absolute least amount of interest in reading docs, reading debug logs, etc. Just my anecdotal experience :) 

Tl;dr i'd prefer to see this feature live in vendor distributions of this library, via a Railtie.

**Comment by anuraaga** [comment]
It sounds like there is a reasonable approach for this for rails, but nothing generic for any ruby library. Given rails' popularity I guess it's OK and probably the best we can do so will go with that. I could verify that a rails initializer can be applied automatically and sends traces without app changes. Thanks for the clarification!
